pkg/cmd/server/origin/handlers.go

```diff
@@ -168,6 +176,98 @@ func namespacingFilter(handler http.Handler, contextMapper kapi.RequestContextMa
 	})
 }
 
+// versionSkewFilter adds a filter that may deny PUT an POST requests since they stomp entire resource objects from skewed
+// oc clients, since we know that those clients will strip unknown fields which can lead to unexpected outcomes
+func (c *MasterConfig) versionSkewFilter(binaryInfo version.Info, handler http.Handler) http.Handler {
+	skewedClientPolicy := c.Options.PolicyConfig.SkewedClientPolicy
+	if skewedClientPolicy == configapi.AllowAll {
+		return handler
+	}
+	seg := strings.SplitN(binaryInfo.GitVersion, "-", 2)
+	serverVersionNumber := seg[0][1:]
+	serverCommit := binaryInfo.GitCommit
+	if len(serverCommit) > 7 {
+		serverCommit = serverCommit[:7]
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if req.Method != "PUT" && req.Method != "POST" {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		userAgent := req.Header.Get("User-Agent")
+		if len(userAgent) == 0 {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		// in pre-1.2 oc clients, the user agent is corrupted when we access kubernetes resources and incorrect versions
+		// are passed.  This means that we can't perform any comparison, so we allow this
+		// for example: oc/v1.2.0 (linux/amd64) kubernetes/bc4550d
+		// that value is passed for a 1.1.3 client when accessing kube resources
+		if strings.Contains(userAgent, "kubernetes/") {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		// user agent is: binary/versionNumber (operating system) openshift/commit
+		// for example: oc/v1.1.3 (linux/amd64) openshift/9b52c41
+		// only match on oc.  This is very restrictive check.  We want to allow by default here.  Clients effectively "opt-in" to
+		// stop stomping things they don't recognize
+		tokens := strings.SplitN(userAgent, "oc/v", 2)
+		if len(tokens) != 2 {
```

Comment:
As mentioned above the point of a skew client check is to give admins a
tool to prevent problems they hit.  Oc isn't the only client - the fact
that we are only worried about it right now is just because it's the first
example, but the actual feature we want is "deny clients based on user
agent".  The selective stripping checks are an optimization for a
particular problem (to be selective), but the more general "deny clients
that report these user agents and give them a clean message" is the feature
we want to offer to admins.

On Wed, Mar 2, 2016 at 5:16 PM, David Eads <notifications@github.com> wrote:

> In pkg/cmd/server/origin/handlers.go
> <https://github.com/openshift/origin/pull/7741#discussion_r54810003>:
>
> > +
> > +		// in pre-1.2 oc clients, the user agent is corrupted when we access kubernetes resources and incorrect versions
> > +		// are passed.  This means that we can't perform any comparison, so we allow this
> > +		// for example: oc/v1.2.0 (linux/amd64) kubernetes/bc4550d
> > +		// that value is passed for a 1.1.3 client when accessing kube resources
> > +		if strings.Contains(userAgent, "kubernetes/") {
> > +			handler.ServeHTTP(w, req)
> > +			return
> > +		}
> > +
> > +		// user agent is: binary/versionNumber (operating system) openshift/commit
> > +		// for example: oc/v1.1.3 (linux/amd64) openshift/9b52c41
> > +		// only match on oc.  This is very restrictive check.  We want to allow by default here.  Clients effectively "opt-in" to
> > +		// stop stomping things they don't recognize
> > +		tokens := strings.SplitN(userAgent, "oc/v", 2)
> > +		if len(tokens) != 2 {
>
> There's no reason to believe they'd submit the same format of user agent.
> The code can be reworked if and when that is required, probably with
> priority regex matching, it's not particularly dense or layered.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7741/files#r54810003>.
>